### PR TITLE
Support generating conversions/deep-copy funcs in multiple packages

### PR DESCRIPTION
**Pull request by uluyol**
This PR tackles the first problem discussed in #12160.

With this PR, we introduce the notion of a "generation unit"—a prefix that defines a set of packages for which conversions/deep-copies may be generated*, assuming that conversions/deep-copies outside of the prefix already exist. This way, we can e.g. generate deep copies for pkg/util.Time in pkg/api, and simple use them as-is in pkg/api/v1. Likewise, this allows for conversions to be separately generated for the experimental and stable api prefixes.
- we still choose which types to generate conversions for by looking at the types defined in the target package that are registered under api.Scheme. The need for a prefix arises because types may use fields with types from other packages like pkg/util.

@smarterclayton @nikhiljindal @wojtek-t

### CODE REVIEW

**Comment by uluyol** [comment]
I believe that this PR would also address openshift's conversion issues.

File: pkg/runtime/conversion_generator.go [hunk 1]
```diff
@@ -102,15 +109,38 @@ func (g *conversionGenerator) GenerateConversionsForType(version string, reflect
...
+	switch t.Kind() {
+	case reflect.Ptr:
+		return g.inGenUnit(t.Elem())
+	case reflect.Struct:
+		return strings.HasPrefix(t.PkgPath(), g.targetUnit)
+	}
+	return true
+}
+
+func (g *conversionGenerator) canConvert(inType, outType reflect.Type) (validConversion bool, err error) {
```
Comment (caesarxuchao) [inline pkg/runtime/conversion_generator.go]: We don't need to name the return value here.

**Comment by wojtek-t** [comment]
I had few comments in the individual commits, but in general the direction looks good to me.

**Comment by caesarxuchao** [comment]
I don't have more comments now. Please address wojtek-t's comments. Thank you!

**Comment by wojtek-t** [comment]
Can you please update your version (after applying comments) and rebase?

**Comment by uluyol** [comment]
Sorry, focusing on getting kubectl changes for experimental done (should have a PR later today) along with just creating duplicate conversions where needed. That should unblock all the PRs that are waiting on experimental.

**Comment by caesarxuchao** [comment]
Superseded by #12411

**Comment by wojtek-t** [comment]
Can we close this PR then?

**Comment by uluyol** [comment]
#12411 doesn't replace this PR because it doesn't solve the duplication issue, but sure we can make a new PR when that comes instead of keeping this around.

**Pull request closed without merging**